### ISSUE TITLE
Tracing enhancements and correlations

### ISSUE THREAD

**Issue by jotak** [kiali/kiali#2776]
Grouping here several tasks of different sizes around tracing & correlation (but each one is self-contained)

- [x] In-graph tracing #2677 (size XL?)
- [x] Add trace-based information in Health #2707 (size L?) (**Closed**)
- [x] Metrics page: more details & links from spans #2775 (size M?)
- [x] Tracing tab in service view #3071
- [x] Tracing tab in workload view #2710
- [X] _Warning check by service #2708_ (**Removed**)
- [ ] Correlation Dashboard #2777 (also related to #1981 and #2708 **closed**) (size XXL?)
- [x] Bucketed display when too many traces/spans #2198 (size L?)
- [x] Traces on workloads/services #3071 , #3048
- [x] Tracing details fix and enhancements #3095
  - [x] Follow-up: Use metrics-based comparison #3293
  - [x] Follow-up: Span filters toolbar #3278
- [ ] Trace mini-graph #3182
- [x] Comparison map per span #3293
- [x] Comparison map per trace #3279
- [ ] Correlate tracing with logging

Also, maintenance work / pre-requisite:
- [x] #2732 (done @jotak)

**Comment by jotak** [comment]
I've added a new task under this epic: #3279 (showing traces heatmap)
Once this is done I want to write a comprehensive blog post to recap all the tracing things done in Kiali since my previous post on that subject [1] which is almost one year old.

But then perhaps it is also time to close this epic? Because it's kind of a never-ending thing, we will keep improving traces anyway (with or without the epic), at the moment I feel it's more of a burden to keep maintaining a sub-tasks list. It would be actually simpler to manage, I think, if we could label tasks with "areas"; like a label "area/tracing" for instance. And for us easier to retrieve where we stand in a specific area in terms of backlog. Thoughts? (@lucasponce @jshaughn )

[1] https://medium.com/kialiproject/tracing-in-kiali-correlation-with-metrics-6e7c1cc67e2a

**Comment by lucasponce** [comment]
> But then perhaps it is also time to close this epic? Because it's kind of a never-ending thing, we will keep improving traces anyway (with or without the epic), at the moment I feel it's more of a burden to keep maintaining a sub-tasks list. It would be actually simpler to manage, I think, if we could label tasks with "areas"; like a label "area/tracing" for instance. And for us easier to retrieve where we stand in a specific area in terms of backlog. Thoughts? (@lucasponce @jshaughn )

Yes, I agree. The Epic makes sense when we want to start a big effort that require some subtasks (and some coordination).

I guess for this area we are in a stage that we can move to more individual/regular issues within an area as other.

Later today I'll prepare the Sprint 49 and I'll revisit this.

**Comment by lucasponce** [comment]
But before to closing, we can discuss on next Scrum with @aljesusg because we have other subtasks there.

**Comment by lucasponce** [comment]
>  at the moment I feel it's more of a burden to keep maintaining a sub-tasks list.

subtasks will be necessary, because those helps to visualize the progress of an epic.
I don't see a real burden if those are necessary, if you compare with other systems JIRA, I'm sure you'd prefer to have a subtask in a github issue :)

**Comment by jotak** [comment]
> But before to closing, we can discuss on next Scrum with @aljesusg because we have other subtasks there.

You're right, sorry, I've got a couple of tasks that I would like to rule out from this epic, and there's also these two from @aljesusg :
- https://github.com/kiali/kiali/issues/2707
- https://github.com/kiali/kiali/issues/2708
About the second one, there's been a lot of discussions and I'm not sure exactly where we stand today, if this is still to be done here, as it crossed the "alerting" story too...

**Comment by lucasponce** [comment]
> About the second one, there's been a lot of discussions and I'm not sure exactly where we stand today, if this is still to be done here, as it crossed the "alerting" story too...

Let's discuss them on next Scrum meeting, I'm fine to refine them and probably move them for better clarity (perhaps we need to create an epic about health config and that can be the context).

**Comment by jotak** [comment]
Hi @aljesusg @lucasponce

There's two tasks currently to add or change options from the traces chart toolbar:

- The one to have percentile-based filtering instead of the current response time slices: https://github.com/kiali/kiali/issues/3380
- The one to introduce spans average duration in chart: https://github.com/kiali/kiali/issues/3399

Perhaps it's the opportunity to redesign the toolbar and make it more alike what we have in other pages? In Graph or Metrics, most display options are grouped inside a single dropdown. Is it the way to go here? (should be discussed with @lwrigh )

I've built this quick mockup to show how it could be like:
![Capture d’écran de 2020-12-08 12-45-19](https://user-images.githubusercontent.com/2153442/101485733-dc8d7f80-395b-11eb-9790-f27a87c26f88.png)

Here I removed the "response code" filter, and decreased a lot the options on query limit (I don't think it's very useful to have so many limits, what do you think?).

About the response code, I'd like to know your thoughts. Currently we can filter only by individual response code (200, 403, 404, etc.). It's an arbitrary list that we have. Ideally, we would certainly prefer to be able to filter with coarser granularity (4xx, 5xx) but Jaeger doesn't allow to query tags using a regex, hence we have this flat-list selection per individual response code. Now the question is: how useful is it? Personally I've never used it, I rely more on the "error" tag when I need to filter for errors, although it's less accurate I find it more interesting to get all the traces with errors at once, not having to iterate through different error codes. So, given my own experience, I'd be in favour of removing this filter, but of course i'd like to hear from other's experiences.

Last thing: I'd like to add a new filter by operation name. Probably it wouldn't fit in the new display dropdown, so would be displayed next to it.

**Comment by lwrigh** [comment]
@jotak Can you attach a screenshot of what the dropdown looks like currently so I can compare the two?

**Comment by jotak** [comment]
@lwrigh sure, here's the current toolbar (with several dropdowns):

![Capture d’écran de 2020-12-08 17-21-59](https://user-images.githubusercontent.com/2153442/101510295-0a81bc80-397a-11eb-889d-7e15e0a1138b.png)

**Comment by lwrigh** [comment]
Thank you for sharing it @jotak! I definitely agree that using the unified Display dropdown is a better option than the current one.

**Comment by lucasponce** [comment]
> Hi @aljesusg @lucasponce
> 
> There's two tasks currently to add or change options from the traces chart toolbar:
> 
> * The one to have percentile-based filtering instead of the current response time slices: #3380
> * The one to introduce spans average duration in chart: #3399
> 
> Perhaps it's the opportunity to redesign the toolbar and make it more alike what we have in other pages? In Graph or Metrics, most display options are grouped inside a single dropdown. Is it the way to go here? (should be discussed with @lwrigh )

+1

> 
> I've built this quick mockup to show how it could be like:
> ![Capture d’écran de 2020-12-08 12-45-19](https://user-images.githubusercontent.com/2153442/101485733-dc8d7f80-395b-11eb-9790-f27a87c26f88.png)
> 
> Here I removed the "response code" filter, and decreased a lot the options on query limit (I don't think it's very useful to have so many limits, what do you think?).
> 
> About the response code, I'd like to know your thoughts. Currently we can filter only by individual response code (200, 403, 404, etc.). It's an arbitrary list that we have. Ideally, we would certainly prefer to be able to filter with coarser granularity (4xx, 5xx) but Jaeger doesn't allow to query tags using a regex, hence we have this flat-list selection per individual response code. Now the question is: how useful is it? Personally I've never used it, I rely more on the "error" tag when I need to filter for errors, although it's less accurate I find it more interesting to get all the traces with errors at once, not having to iterate through different error codes. So, given my own experience, I'd be in favour of removing this filter, but of course i'd like to hear from other's experiences.
> 

This filter comes from the first days of this implementation, so I'm +1 to just separate "errors" vs "all"

> Last thing: I'd like to add a new filter by operation name. Probably it wouldn't fit in the new display dropdown, so would be displayed next to it.

About this I may have more ideas like adding possibility to filter by a specific tag as well, but I guess that's a parallel discussion.

To my, just to settle, the purpose of this timeline component should be to spot user in the interesting group of traces to allow to move to the next step which should be click on a specific trace for details examination.

**Comment by lucasponce** [comment]
There are a couple of open issues that may require more discussion and most likely can be part of a new future Epic.
But this one is mostly complete, so I'm going to marked as done.
